File: Data Structures & Algorithms/invert-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return

        if root.right == None and root.left == None:
            return root

        elif root.right == None:
            self.invertTree(root.left)
            root.right = root.left
            root.left = None

        elif root.left == None:
            self.invertTree(root.right)
            root.left = root.right
            root.right = None
        else:
            self.invertTree(root.left)
            self.invertTree(root.right)
            temp = root.left
            root.left = root.right
            root.right = temp

        return root

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfscheck(root, -1001, 1001);
    }

    bool dfscheck(TreeNode* root, long small, long big) {
            if (!root) {
                return true;
            }
            
            if  (!(small < root->val and root->val < big)) {
                return false;
            }

            return dfscheck(root->left, small, root->val) and dfscheck(root->right, root->val, big);
        }
};
